src/VecSim/containers/data_blocks_container.cpp

```diff
@@ -4,48 +4,40 @@ DataBlocksContainer::DataBlocksContainer(size_t blockSize, size_t elementBytesCo
                                          std::shared_ptr<VecSimAllocator> allocator,
                                          unsigned char _alignment)
     : VecsimBaseObject(allocator), RawDataContainer(), element_bytes_count(elementBytesCount),
-      len(0), blocks(allocator), block_size(blockSize), alignment(_alignment) {}
+      element_count(0), blocks(allocator), block_size(blockSize), alignment(_alignment) {}
 
 DataBlocksContainer::~DataBlocksContainer() = default;
 
-size_t DataBlocksContainer::size() const { return len; }
+size_t DataBlocksContainer::size() const { return element_count; }
 
-RawDataContainer_Status DataBlocksContainer::appendElement(const void *element) {
-    if (len % block_size == 0) {
-        blocks.push_back(DataBlock(this->block_size, this->element_bytes_count, this->allocator,
-                                   this->alignment));
+RawDataContainer_Status DataBlocksContainer::addElement(const void *element, size_t id) {
+    assert(id == element_count); // we can only append new elements
+    if (element_count % block_size == 0) {
+        blocks.emplace_back(this->block_size, this->element_bytes_count, this->allocator,
+                            this->alignment);
     }
     blocks.back().addElement(element);
-    len++;
+    element_count++;
     return RAW_DATA_CONTAINER_OK;
 }
 
 const char *DataBlocksContainer::getElement(size_t id) const {
-    if (id >= len) {
-        return nullptr;
-    }
+    assert(id < element_count);
     return blocks.at(id / this->block_size).getElement(id % this->block_size);
 }
 
 RawDataContainer_Status DataBlocksContainer::removeElement(size_t id) {
-    if (id >= len) {
-        return RAW_DATA_CONTAINER_ID_NOT_EXIST;
-    }
-    if (id < len - 1) {
-        return RAW_DATA_CONTAINER_ERR; // only the last element can be removed
-    }
+    assert(id == element_count - 1); // only the last element can be removed
```

Comment:
Why not implement the remove as an internal swap with the last element?

Reply:
The ID is an external handle. If we would swap, then you would loose the original mapping (or you would have to maintain another mapping, which is not desired in this case...)